modules/PanelApp.py

```diff
@@ -0,0 +1,227 @@
+                )
+            )
+            .withColumn('phenotype', regexp_replace(col('phenotype'), f'({self.OTHER_RE})', ''))
+
+            # Extract OMIM identifiers and remove them from the phenotype string.
+            .withColumn('omim_id', regexp_extract(col('phenotype'), self.OMIM_RE, 2))
+            .withColumn('omim', when(col('omim_id') != '', concat(lit('OMIM:'), col('omim_id'))))
+            .withColumn('phenotype', regexp_replace(col('phenotype'), f'({self.OMIM_RE})', ''))
+
+            # Choose one of the ontology identifiers, keeping OMIM as a priority.
+            .withColumn('diseaseFromSourceId', when(col('omim').isNotNull(), col('omim')).otherwise(col('ontology')))
+            .drop('ontology_namespace', 'ontology_id', 'ontology', 'omim_id', 'omim')
+
+            # Clean up the final split phenotypes.
+            .withColumn('phenotype', regexp_replace(col('phenotype'), r'\(\)', ''))
+            .withColumn('phenotype', trim(col('phenotype')))
+            .withColumn('phenotype', when(col('phenotype') != '', col('phenotype')))
+
+            # Remove empty or low quality records.
+            .filter(
+                ~(
+                    # There is neither a phenotype string nor an ontology identifier.
+                    ((col('phenotype').isNull()) & (col('diseaseFromSourceId').isNull())) |
+                    # The name of the phenotype string starts with a question mark, indicating low quality.
+                    col('phenotype').startswith('?')
+                )
+            )
+            .persist()
+        )
+
+        # Fetch and join literature references.
+        all_panel_ids = panelapp_df.select('Panel Id').toPandas()['Panel Id'].unique()
+        literature_references = self.fetch_literature_references(all_panel_ids)
+        panelapp_df = panelapp_df.join(literature_references, on=['Panel Id', 'Symbol'], how='left')
+
+        # Populate final evidence string structure.
+        panelapp_df = (
+            panelapp_df
+            # allelicRequirements requires a list, but we always only have one value from PanelApp.
+            .withColumn(
+                'allelicRequirements',
+                when(col('Mode of inheritance').isNotNull(), array(col('Mode of inheritance')))
+            )
+            .drop('Mode of inheritance')
+
+            .withColumnRenamed('List', 'confidence')
+            .withColumn('datasourceId', lit('genomics_england'))
+            .withColumn('datatypeId', lit('genetic_literature'))
+            .withColumnRenamed('phenotype', 'diseaseFromSource')
+            # diseaseFromSourceId populated above
+            # literature populated above
+            .withColumnRenamed('Panel Id', 'studyId')
+            .withColumnRenamed('Panel Name', 'studyOverview')
+            .withColumnRenamed('Symbol', 'targetFromSourceId')
+        )
+
+        # Save data.
+        with tempfile.TemporaryDirectory() as tmp_dir_name:
```

Comment:
I like this! Writing a temporary file, doing a check and then renaming. Once we are touching the other parsers, we can apply this practice.

Reply:
The main utility of this code fragment is making sure that the output is indeed one single file, and this is what the majority of it (7 out of 8 lines) is doing.

By default, if we just call `.coalesce(1).write`, it will create a directory with one, randomly named, gzipped JSON inside, which is not always convenient. By doing the chunk discovery and renaming, we are making sure that there is only one file and it is consistently named.

The remaining 1 line, the assert, is just there as a precaution for a very unlikely scenario. It's incredibly cheap to write (a few seconds of my time), but can trigger in certain situations. For example, if we ever run the parsers on network filesystems (NFS, fuse mounts), it's a very useful check to have.